docs/DifferentiableFunctions.md

```diff
@@ -0,0 +1,444 @@
+print(x.valueWithGradient(in: square)) // (value: 9.0, gradient: 6.0)
+```
+
+Here's a list of differentiation APIs provided by the standard library:
+
+| Differentiation APIs  | Description                                                                                                                                                      |
+|-----------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| [`valueWithPullback(at:in:)`] <br/> [`valueWithPullback(at:_:in:)`] | Returns original result and backpropagation function. <br/> **Important note: this is the core differentiation API. All other APIs are defined in terms of `valueWithPullback`.** |
+| [`pullback(at:in:)`] <br/> [`pullback(at:_:in:)`] | Returns backpropagation function. |
+| [`gradient(at:in:)`] <br/> [`gradient(at:_:in:)`] | Returns partial derivatives with respect to arguments. |
+| [`valueWithGradient(at:in:)`] <br/> [`valueWithGradient(at:_:in:)`] | Returns original result and partial derivatives with respect to arguments. |
+| [`gradient(of:)`] <br/> [`gradient(of:)` (arity 2)] | Returns gradient function. |
+| [`valueWithGradient(of:)`] <br/> [`valueWithGradient(of:)` (arity 2)] | Returns gradient function. |
+
+### Constructing `@differentiable` functions
+
+A value with a non-differentiable function type can be implicitly converted to
+one with a corresponding `@differentiable` function type. In fact, this is what
+happened in the example above:
+
+```swift
+// `square` has type `(Float) -> Float`.
+func square(_ x: Float) -> Float {
+    return x * x
+}
+let x: Float = 3.0
+
+// The second argument of `gradient(at:in:)` has type `@differentiable (Float) -> Float`.
+// When `square` is passed to `gradient(at:in:)`, it is implicitly converted to a value with the
+// `@differentiable` type.
+print(gradient(at: x, in: square)) // 6.0
+```
+
+The implicit conversion from a value with type `(T) -> U` to a value with type
+`@differentiable (T) -> U` actually triggers differentiation by the compiler.
+Thus, differentiation is type-driven.
+
+If differentiation succeeds, the `@differentiable (T) -> U` value is
+constructed. If differentiation fails, the compiler emits a compile-time error:
+
+```swift
+let add: @differentiable (Float, Float) -> Float = { x, y in
+    // The `Int` initializer call below is non-differentiable.
+    Float(Int(x + y))
+}
+```
+
+```console
+test.swift:1:52: error: function is not differentiable
+let add: @differentiable (Float, Float) -> Float = { x, y in
+                                                   ^~~~~~~~~
+test.swift:2:11: note: cannot differentiate through a non-differentiable result; do you want to add '.withoutDerivative()'?
+    Float(Int(x + y))
+          ^
+```
+
+There are a few reasons why differentiation can fail:
+* The function to differentiate contains non-differentiable computation along
```

Comment:
Earlier Conal Elliott on Twitter made a good point about the distinction between “non-differentiable functions” (mathematically non differentiable functions) and “functions whose derivatives are not computable” (opaque functions). Since our language doesn’t know math, we don’t really know what functions are really nondifferentiable in the mathematical sense. So we’ve been using “cannot differentiate xxx” in our diagnostics instead of “xxx is nondifferentiable”. We should probably use the same wording here instead of “nondifferentiable”.

Reply:
How about:
`* The function to differentiate contains dataflow, from parameters to result, that cannot be differentiated.`

Reply:
“Data flow” or “dataflow” is a compiler term IMO. The user understands “code” or “computation”.

Reply:
How about:
```
* The function to differentiate contains computation, from parameters to result, that cannot be differentiated.
```

And also, how about adding another point for opaque functions?
```
* The function to differentiate is opaque, i.e. it is a function parameter with a non-`@differentiable` function type.
```

Reply:
I think both these are good. Any inability to differentiate something always comes from encountering an opaque, non-@differentiable function.

Reply:
Added both points.